### PR TITLE
make sure the CachedAssetReader/CachedAssetWriter dont throw synchronously

### PR DESCRIPTION
**Pull request by jakemac53**
The easier solution here would be to just mark all the methods as `async`, but we need to deal with the caches in a synchronous fashion. Those same caches can throw however, so wrapping everything in a try/catch seems to be the simplest solution.

### CODE REVIEW

File: lib/src/asset/cache.dart [hunk 1]
```diff
@@ -58,27 +58,41 @@ class CachedAssetReader extends AssetReader {
...
+
+      return _pendingHasInputChecks.putIfAbsent(id, () async {
+        try {
+          return await _reader.hasInput(id);
+        } finally {
+          _pendingHasInputChecks.remove(id);
+        }
+      });
+    } catch (e, s) {
+      return new Future.error(e, s);
```
Comment (kevmoo) [inline lib/src/asset/cache.dart]: Or just make it `async`?

File: lib/src/asset/cache.dart [hunk 2]
```diff
@@ -58,27 +58,41 @@ class CachedAssetReader extends AssetReader {
...
+          var content = await _reader.readAsString(id, encoding: encoding);
+          _cache.put(new Asset(id, content));
+          return content;
+        } finally {
+          // Make sure we always remove the pending read
+          _pendingReads.remove(id);
+        }
+      });
+    } catch(e, s) {
+      return new Future.error(e, s);
```
Comment (kevmoo) [inline lib/src/asset/cache.dart]: Or just make it `async`?

**Comment by kevmoo** [comment]
Yeah – why not make these methods `async`? Then you can get rid of all of the explicit `Future` creation

**Comment by jakemac53** [comment]
Because of the caches, which need to be updated synchronously. For instance given the code:

```dart
Future.wait(assetIds.map((assetId) => reader.readAsString(assetId)));
``

All of those calls are dispatched synchronously, so the `_pendingReads` map needs to be updated synchronously.

File: lib/src/asset/cache.dart [hunk 3]
```diff
@@ -58,27 +58,41 @@ class CachedAssetReader extends AssetReader {
 
   @override
   Future<bool> hasInput(AssetId id) {
```
Comment (kevmoo) [inline lib/src/asset/cache.dart]: Add comment as to why this is not `async` – it makes sense now, but docs would help.

**Comment by kevmoo** [comment]
LGTM – just add comments as to why we don't use `async` for future maintainers.

**Pull request merged**